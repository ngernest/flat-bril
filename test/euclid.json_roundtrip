{
  "functions": [
    {
      "args": [],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 23789216
        },
        {
          "args": [
            "v0"
          ],
          "dest": "x",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1748698766
        },
        {
          "args": [
            "v1"
          ],
          "dest": "y",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "f",
          "funcs": [
            "gcd"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "f"
          ],
          "dest": "f",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "f"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v5",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "r",
          "type": "int"
        },
        {
          "name": "s",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "r"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "r"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "s"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "s"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v6"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v0"
          ],
          "dest": "zero",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "zero"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "dest": "cond_temp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "cond_temp"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "not",
          "type": "bool"
        },
        {
          "args": [
            "v4"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "label": "for.cond.5"
        },
        {
          "args": [
            "cond"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [
            "for.body.5",
            "for.end.5"
          ],
          "op": "br"
        },
        {
          "label": "for.body.5"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "dest": "temp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v10"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "temp"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "zero"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v14"
          ],
          "dest": "cond_temp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "cond_temp"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "not",
          "type": "bool"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.5"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.5"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v16"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "gcd",
      "type": "int"
    }
  ]
}
