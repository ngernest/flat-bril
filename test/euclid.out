@main
	op: const		dest: v0		type: int		value: 23789216	
	op: id   		dest: x 		type: int		args: ["v0"]	
	op: const		dest: v1		type: int		value: 1748698766	
	op: id   		dest: y 		type: int		args: ["v1"]	
	op: id   		dest: v2		type: int		args: ["x"]	
	op: id   		dest: v3		type: int		args: ["y"]	
	op: call 		dest: f 		type: int		args: ["v2", "v3"]		funcs: ["gcd"]
	op: id   		dest: f 		type: int		args: ["f"]	
	op: id   		dest: v4		type: int		args: ["f"]	
	op: print		args: ["v4"]	
	op: const		dest: v5		type: int		value: 0	
@mod
	op: id   		dest: v0		type: int		args: ["r"]	
	op: id   		dest: v1		type: int		args: ["r"]	
	op: id   		dest: v2		type: int		args: ["s"]	
	op: div  		dest: v3		type: int		args: ["v1", "v2"]	
	op: id   		dest: v4		type: int		args: ["s"]	
	op: mul  		dest: v5		type: int		args: ["v3", "v4"]	
	op: sub  		dest: v6		type: int		args: ["v0", "v5"]	
	op: id   		dest: result		type: int		args: ["v6"]	
	op: id   		dest: v7		type: int		args: ["result"]	
	op: ret  		args: ["v7"]	
@gcd
	op: const		dest: v0		type: int		value: 0	
	op: id   		dest: zero		type: int		args: ["v0"]	
	op: id   		dest: v1		type: int		args: ["b"]	
	op: id   		dest: v2		type: int		args: ["zero"]	
	op: eq   		dest: v3		type: bool		args: ["v1", "v2"]	
	op: id   		dest: cond_temp		type: bool		args: ["v3"]	
	op: not  		dest: v4		type: bool		args: ["cond_temp"]	
	op: id   		dest: cond		type: bool		args: ["v4"]	
.for.cond.5
	op: id   		dest: v6		type: bool		args: ["cond"]	
	op: br   		args: ["v6"]		labels: ["for.body.5", "for.end.5"]
.for.body.5
	op: id   		dest: v7		type: int		args: ["b"]	
	op: id   		dest: temp		type: int		args: ["v7"]	
	op: id   		dest: v8		type: int		args: ["a"]	
	op: id   		dest: v9		type: int		args: ["b"]	
	op: call 		dest: v10		type: int		args: ["v8", "v9"]		funcs: ["mod"]
	op: id   		dest: b 		type: int		args: ["v10"]	
	op: id   		dest: v11		type: int		args: ["temp"]	
	op: id   		dest: a 		type: int		args: ["v11"]	
	op: id   		dest: v12		type: int		args: ["b"]	
	op: id   		dest: v13		type: int		args: ["zero"]	
	op: eq   		dest: v14		type: bool		args: ["v12", "v13"]	
	op: id   		dest: cond_temp		type: bool		args: ["v14"]	
	op: not  		dest: v15		type: bool		args: ["cond_temp"]	
	op: id   		dest: cond		type: bool		args: ["v15"]	
	op: jmp  		labels: ["for.cond.5"]
.for.end.5
	op: id   		dest: v16		type: int		args: ["a"]	
	op: ret  		args: ["v16"]	
